Windows EC2 instance detection

ogaca-dd opened the pull request:
### What does this PR do?

On windows under EC2, hostname starts with `ec2amaz-`. 
If the system hostname starts with "ec2amaz-" we use the instance id instead of the hostname.

`agent status` output is:

Before

```
 Hostnames
  =========
    ec2-hostname: ip-172-29-3-76.ec2.internal
    hostname: EC2AMAZ-V68JKHF
    instance-id: i-0fc144c877ac77810
    socket-fqdn: EC2AMAZ-V68JKHF
    socket-hostname: EC2AMAZ-V68JKHF
    hostname provider: os
```

After

```
 Hostnames
  =========
    ec2-hostname: ip-172-29-5-33.ec2.internal
    hostname: i-017091562d278613e
    instance-id: i-017091562d278613e
    socket-fqdn: EC2AMAZ-OUBBQHG
    socket-hostname: EC2AMAZ-OUBBQHG
    hostname provider: aws
```

ogaca-dd committed:
Handle Amazon Windows EC2 instances detection

ogaca-dd committed:
Add config flag.

ogaca-dd committed:
Add release note

ogaca-dd committed:
Remove duplicated release note

ogaca-dd committed:
Fix lint error.

albertvaka commented on releasenotes/notes/detect-win-ec2-instance-3e90f6ad7ae03683.yaml:
```diff
@@ -0,0 +1,11 @@
+#
+# Each section note must be formatted as reStructuredText.
+---
+fixes:
```
This is a breaking change, so you might want to add an `upgrade` note.

ogaca-dd replied:
Updated

olivielpeau replied:
note: given the way the change is implemented, this is _not_ a breaking change

ogaca-dd committed:
Add upgrade section to release note.

ogaca-dd committed:
Merge branch 'master' of github.com:DataDog/datadog-agent into ogaca/Win32Ec2Instance

albertvaka commented on releasenotes/notes/detect-win-ec2-instance-3e90f6ad7ae03683.yaml:
```diff
@@ -0,0 +1,14 @@
+---
+upgrade:
+  - |
+    An hostname starting with ``EC2AMAZ-`` has its canonical hostname set to the instance id.
```
```suggestion
    Hostnames starting with ``EC2AMAZ-`` will now have its canonical hostname set as instance id.
    You might need to update these hostnames in your dashboards and alarms accordingly.
```

It's good to provide some actionable upgrade path to make the 'upgrade' notes less scary.

ogaca-dd committed:
Update releasenotes/notes/detect-win-ec2-instance-3e90f6ad7ae03683.yaml

Co-Authored-By: Albert Vaca <albert.vaca@datadoghq.com>

ogaca-dd committed:
Merge branch 'master' of github.com:DataDog/datadog-agent into ogaca/Win32Ec2Instance

albertvaka commented on pkg/config/config.go:
```diff
@@ -326,6 +326,9 @@ func initConfig(config Config) {
 	config.BindEnvAndSetDefault("metadata_endpoints_max_hostname_size", 255)
 
+	// EC2
+	config.BindEnvAndSetDefault("ec2.use_windows_prefix_detection", false)
```
@olivielpeau shouldn't we set this to true, so new users get the correct behaviour?

olivielpeau replied:
unfortunately, if we set this to true, not only will new users get the "better" behavior that this PR implements, but existing users will also get the better behavior when they upgrade (because they most likely won't have the config key defined in their config).

Ideally, we should add a deprecation warning log entry when the option is still set to `false`, and the Agent detects that the resolved hostname would change if the option were set to `true`. That way, users would have a way to know this option exists and that they should set it to `true` to get the better behavior (see what we've documented here: https://docs.datadoghq.com/agent/faq/how-datadog-agent-determines-the-hostname/?tab=agentv6#agent-versions). *Or* we find a way to handle the change gracefully in the intake/backend (tricky).

Still unclear what the path is to then actually make the breaking change of switching the value to `true` by default...

albertvaka replied:
Users upgrading should read the `upgrade` section of the changelog. That's the reason we have that section, so we can make breaking changes when it's needed. I respect your decision if you want to be conservative, but I think it leads to technical debt.

olivielpeau replied:
From past experience, a lot of users do not read the `upgrade` section of the changelog between minor releases of the Agent. For changes that we know will break a lot of users' monitors (for example), I'd advise caution.

You're right, this leads to more technical debt. Making it easier for us to make "small" breaking changes to the Agent is a topic we'll have to tackle.

ogaca-dd replied:
I did the following test:
 - create a fresh instance of a Windows EC2
- install the agent with the default config `ec2.use_windows_prefix_detection", false`
- Run the agent

I got 
![Screen Shot 2019-10-09 at 2 39 03 PM](https://user-images.githubusercontent.com/52180542/66482603-9d3b7880-eaa3-11e9-88d8-1ecbba2bfb7e.png)

Then I update `ec2.use_windows_prefix_detection` at true.
`agent status` output is:
```
 Hostnames
  =========
    ec2-hostname: ip-172-29-17-9.ec2.internal
    hostname: i-055d1f54dd99ce881
    instance-id: i-055d1f54dd99ce881
    socket-fqdn: EC2AMAZ-0I0CNQ5
    socket-hostname: EC2AMAZ-0I0CNQ5
```
On the web app we have:
![Screen Shot 2019-10-09 at 2 49 09 PM](https://user-images.githubusercontent.com/52180542/66482798-04592d00-eaa4-11e9-8ab2-46388acbbcf2.png)

It does not seem to be a breaking changes at least on the host detailed. Did I miss something?

albertvaka replied:
It's not breaking with the default == false, so it's good to merge. I was making a point in favor of making it default to true, but no need to hold the PR for this.

olivielpeau replied:
nit about the name of the option: I think we should use `_` instead of `.` for consistency with other similar options

olivielpeau commented on releasenotes/notes/detect-win-ec2-instance-3e90f6ad7ae03683.yaml:
```diff
@@ -0,0 +1,15 @@
+upgrade:
+  - |
+    Hostnames starting with ``EC2AMAZ-`` will now have their canonical hostname set as instance id.
+    You might need to update these hostnames in your dashboards and alarms accordingly.
```
DD terminology: `alarms` --> `monitors`

ogaca-dd replied:
upgrade comment removed

olivielpeau commented on releasenotes/notes/detect-win-ec2-instance-3e90f6ad7ae03683.yaml:
```diff
@@ -0,0 +1,15 @@
+---
+upgrade:
+  - |
+    Hostnames starting with ``EC2AMAZ-`` will now have their canonical hostname set as instance id.
```
This change is specific to AWS, let's mention that here

ogaca-dd replied:
upgrade comment removed

ogaca-dd committed:
Remove upgrade section in the changelog.

ogaca-dd commented:
Here is the current behavior

**For a new host.**

Case1
--------
- set `ec2.use_windows_prefix_detection` at true
- start the agent
![Screen Shot 2019-10-10 at 3 06 04 PM](https://user-images.githubusercontent.com/52180542/66571423-8c583900-eb6f-11e9-853d-59c1788ed930.png)
=> the instance-id is used as canonical instance
If we set `ec2.use_windows_prefix_detection` at false, we have another host. Metrics are not sent to the previous host.
![Screen Shot 2019-10-10 at 3 10 18 PM](https://user-images.githubusercontent.com/52180542/66571797-2f10b780-eb70-11e9-8331-9e1d218844d1.png)

Case2
--------
- set `ec2.use_windows_prefix_detection` at false
- start the agent
![Screen Shot 2019-10-10 at 3 17 29 PM](https://user-images.githubusercontent.com/52180542/66572367-1e147600-eb71-11e9-8506-ca3c36f84e50.png)
If we set `ec2.use_windows_prefix_detection` at true, nothing chages. Metrics are still sent to the host.

The 2 cases exist because in case1, `EC2AMAZ-J9U50S5` is not an alias for the host.

olivielpeau commented on releasenotes/notes/detect-win-ec2-instance-3e90f6ad7ae03683.yaml:
```diff
@@ -0,0 +1,11 @@
+---
+fixes:
+  - |
+    Handle Amazon Windows EC2 instances detection which have default hostnames starting with ``EC2AMAZ-`` on Windows.
```
let's remove the release note altogether until we can fully predict the impact of enabling the config option

olivielpeau reviewed (commented):
a couple of nits. Once addressed I think we can merge this for 6.15, and work on evaluating the impact of toggling the option to true by default for 6.16 (not sure at this point if that's going to break existing setups or not).

ogaca-dd committed:
Small fixes

- Remove changelog
- ec2.use_windows_prefix_detection to ec2_use_windows_prefix_detection

olivielpeau reviewed (approved):
LGTM for 6.15 👍

Let's follow up for 6.16 on whether we can enable the config option by default, and whether a change is needed in the backend as well
